codegen/value.go

```diff
@@ -52,9 +54,71 @@ func NewRegister(ctx context.Context) Register {
 	}
 }
 
-func (r *register) Set(iface interface{}) (err error) {
-	r.Value, err = r.ctor(iface)
-	return err
+func (r *register) Set(iface interface{}) error {
+	return r.SetAsync(func(Value) (Value, error) {
+		return r.ctor(iface)
+	})
+}
+
+func (r *register) SetAsync(f func(Value) (Value, error)) error {
+	var prev chan Value
+	if r.last == nil {
+		prev = make(chan Value, 1)
+		prev <- r.Value
+	} else {
+		prev = r.last
+	}
+
+	ret := make(chan Value)
+	r.last = ret
+
+	go func() {
+		defer close(prev)
```

Comment:
Why do you need to close `prev`? Isn't the line below the last thing that will ever read from it?

Reply:
Huh, I always thought you need to eventually close every channel in go. TIL

Reply:
It's a common misunderstanding. Channels get garbage collected like other variables in Go. Closing it just causes reads to return the zero value immediately once the buffer is empty, and reads with the `x, ok := <-chan` syntax return false for `ok`.